contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -2876,16 +2876,148 @@ add_entry_to_bbf_schema(const char *schema_name,
 	CommandCounterIncrement();
 }
 
+/*get permission entry from catalog*/
+int16
+get_bbf_schema_privilege(const char *schema_name,
+ 							const char *object_name,
+							const char *grantee)
+{
+	Relation	bbf_schema_rel;
+	HeapTuple	tuple_bbf_schema;
+	ScanKeyData	key[4];
+	TableScanDesc	scan;
+	int16	dbid = get_cur_db_id();
+	int16 permission = 0;
+
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+									AccessShareLock);
+	ScanKeyInit(&key[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyInit(&key[1],
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(schema_name));
+	ScanKeyInit(&key[2],
+				Anum_bbf_schema_perms_object_name,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(object_name));
+	ScanKeyInit(&key[3],
+				Anum_bbf_schema_perms_grantee,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(grantee));
+
+	scan = table_beginscan_catalog(bbf_schema_rel, 4, key);
+	tuple_bbf_schema = heap_getnext(scan, ForwardScanDirection);
+	if (HeapTupleIsValid(tuple_bbf_schema))
+	{
+		Form_bbf_schema_perms bbf_schema = (Form_bbf_schema_perms) GETSTRUCT(tuple_bbf_schema);
+		permission = bbf_schema->permission;
+	}
+
+	table_endscan(scan);
+	table_close(bbf_schema_rel, AccessShareLock);
+	return permission;
+}
+
+/*Update catalog entry*/
+void
+update_bbf_schema_entry(const char *schema_name,
+							const char *object_name,
+							int16 new_priv,
+							int16 old_priv,
+							const char *grantee,
+							const char *object_type,
+							bool is_grant)
+{
+	Relation	bbf_schema_rel;
+	HeapTuple	tuple_bbf_schema;
+	TupleDesc	bbf_schema_dsc;
+	HeapTuple	new_tuple;
+	ScanKeyData key[5];
+	TableScanDesc scan;
+	int16	dbid = get_cur_db_id();
+	int16 permission = 0;
+	Datum		new_record_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS];
+	bool		new_record_nulls_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS];
+	bool		new_record_repl_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS];
+
+
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+									RowExclusiveLock);
```

Comment:
You don't need to lock the table here if permission = 0. So, let's delay this till permission == 0 check.